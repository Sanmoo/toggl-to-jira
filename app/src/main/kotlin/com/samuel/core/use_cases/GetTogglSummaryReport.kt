package com.samuel.core.use_cases

import com.samuel.core.entities.WorklogSummary
import com.samuel.core.use_cases.ports.TogglClient
import java.time.ZonedDateTime

class GetTogglSummaryReport(private val togglClient: TogglClient) {
    data class Output(val list: List<WorklogSummaryOutput>)
    data class WorklogSummaryOutput(val title: String, val timeInSeconds: Long, val start: ZonedDateTime) {
        companion object {
            fun fromEntity(l: WorklogSummary): WorklogSummaryOutput {
                return WorklogSummaryOutput(title = l.title, timeInSeconds = l.timeInSeconds, start = l.start)
            }
        }
    }

    fun get(day: ZonedDateTime): Output {
        val worklogSummariesByDate = togglClient.getWorklogSummariesByDate(day)
        return Output(list = worklogSummariesByDate.map { WorklogSummaryOutput.fromEntity(it) })
    }
}