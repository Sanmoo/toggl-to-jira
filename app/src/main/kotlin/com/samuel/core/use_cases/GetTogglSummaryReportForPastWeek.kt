package com.samuel.core.use_cases

import java.time.DayOfWeek
import java.time.ZonedDateTime

class GetTogglSummaryReportForPastWeek(
    private val getTogglSummaryReport: GetTogglSummaryReport
) {
    data class Input(val weeksOffsetFromToday: Int)
    data class Output(val data: Map<ZonedDateTime, GetTogglSummaryReport.Output>)

    fun get(input: Input): Output {
        val beginningOfWeek = ZonedDateTime.now().minusWeeks(input.weeksOffsetFromToday.toLong()).with(DayOfWeek.MONDAY)
        val resultMap = HashMap<ZonedDateTime, GetTogglSummaryReport.Output>()
        (0..6).forEach {
            val dateToQuery = beginningOfWeek.plusDays(it.toLong())
            resultMap[dateToQuery] = getTogglSummaryReport.get(dateToQuery)
        }

        return Output(data = resultMap.toSortedMap())
    }
}