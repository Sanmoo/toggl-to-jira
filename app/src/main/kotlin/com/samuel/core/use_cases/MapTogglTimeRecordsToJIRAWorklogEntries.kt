package com.samuel.core.use_cases

import com.samuel.core.entities.timeSpentAsString
import java.time.ZonedDateTime
import java.time.format.DateTimeFormatter

class MapTogglTimeRecordsToJIRAWorklogEntries {
    data class TogglTimeRecordInput(val title: String, val timeInSeconds: Long, val start: ZonedDateTime)
    data class JiraWorklogEntryOutput(
        val issueKey: String,
        val amountInSeconds: Int,
        val description: String?,
        val start: ZonedDateTime
    )

    fun map(input: List<TogglTimeRecordInput>): List<JiraWorklogEntryOutput> {
        val output = mutableListOf<JiraWorklogEntryOutput>()
        input.forEach { item ->
            val regex = "([A-Z]+-\\d+):?\\s*(.*)".toRegex(RegexOption.IGNORE_CASE)
            val matchResult = regex.matchEntire(item.title)
            if (matchResult == null) {
                println(
                    "Log entry \"${item.title}\", which took ${
                        timeSpentAsString(item.timeInSeconds.toInt())
                    } and started at ${item.start.format(DateTimeFormatter.ofPattern("dd/MM/yyyy HH:mm:ss"))}, is not properly formatted to be sent to JIRA. So skipping..."
                )
                return@forEach
            }
            val groupValues = matchResult.groupValues
            val description = (if (groupValues.size > 2) groupValues[2] else null)?.let { it.ifEmpty { null } }
            output.add(
                JiraWorklogEntryOutput(
                    groupValues[1],
                    item.timeInSeconds.toInt(),
                    description,
                    item.start
                )
            )
        }

        return output
    }
}