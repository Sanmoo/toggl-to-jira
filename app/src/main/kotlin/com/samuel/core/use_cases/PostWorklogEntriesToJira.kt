package com.samuel.core.use_cases

import com.samuel.core.entities.JiraWorklogEntry
import com.samuel.core.use_cases.MapTogglTimeRecordsToJIRAWorklogEntries.JiraWorklogEntryOutput
import com.samuel.core.use_cases.ports.JiraClient

class PostWorklogEntriesToJira(
    private val jiraClient: JiraClient
) {
    fun post(input: List<JiraWorklogEntryOutput>) {
        input.forEach {
            val entry = JiraWorklogEntry(
                issueKey = it.issueKey,
                amountInSeconds = it.amountInSeconds,
                description = it.description,
                start = it.start
            )

            if (jiraClient.isSimilarEntryAlreadyRegistered(entry)) {
                println("Skipping registry of similar entry in JIRA...")
                return@forEach
            }

            jiraClient.logWork(entry)
        }
    }
}