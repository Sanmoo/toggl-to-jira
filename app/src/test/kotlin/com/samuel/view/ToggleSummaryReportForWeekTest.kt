package com.samuel.view

import au.com.origin.snapshots.Expect
import au.com.origin.snapshots.junit5.SnapshotExtension
import com.samuel.core.use_cases.GetTogglSummaryReport
import com.samuel.core.use_cases.GetTogglSummaryReportForPastWeek
import org.junit.jupiter.api.Test
import org.junit.jupiter.api.extension.ExtendWith
import java.time.ZoneId
import java.time.ZonedDateTime

@ExtendWith(SnapshotExtension::class)
internal class ToggleSummaryReportForWeekTest {
    private lateinit var expect: Expect

    @Test
    fun `render report correctly`() {
        val day1 = ZonedDateTime.of(2022, 3, 14, 8, 0, 0, 0, ZoneId.systemDefault())
        val day2 = day1.plusDays(1)
        val day3 = day2.plusDays(1)
        val sut = ToggleSummaryReportForWeek(
            input = GetTogglSummaryReportForPastWeek.Output(
                data = mapOf(
                    day1 to GetTogglSummaryReport.Output(
                        list = listOf(
                            GetTogglSummaryReport.WorklogSummaryOutput(title = "Soy un log amigo", timeInSeconds = 328489, start = day1),
                            GetTogglSummaryReport.WorklogSummaryOutput(title = "Soy un otro log", timeInSeconds = 3289, start = day1)
                        )
                    ),
                    day2 to GetTogglSummaryReport.Output(
                        list = listOf(
                            GetTogglSummaryReport.WorklogSummaryOutput(title = "JIRA-123: Done this and that", timeInSeconds = 28489, start = day2),
                            GetTogglSummaryReport.WorklogSummaryOutput(title = "Done this and that and that", timeInSeconds = 32890, start = day2)
                        )
                    ),
                    day3 to GetTogglSummaryReport.Output(
                        list = listOf(
                            GetTogglSummaryReport.WorklogSummaryOutput(title = "JIRA-123: Done this and that", timeInSeconds = 28489, start = day3),
                            GetTogglSummaryReport.WorklogSummaryOutput(title = "JIRA-432: Dooooooone this and that", timeInSeconds = 35890, start = day3),
                            GetTogglSummaryReport.WorklogSummaryOutput(title = "JIRA-1500: Dooooooone this and that", timeInSeconds = 65890, start = day3)
                        )
                    )
                )
            )
        )

        expect.toMatchSnapshot(sut.render())
    }
}