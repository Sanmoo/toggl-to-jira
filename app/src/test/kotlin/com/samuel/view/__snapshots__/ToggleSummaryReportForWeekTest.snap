com.samuel.view.ToggleSummaryReportForWeekTest.render report correctly=[
14/03/2022
----------

Soy un log amigo (91h 14m 49s)
Soy un otro log (54m 49s)

15/03/2022
----------

JIRA-123: Done this and that (7h 54m 49s)
Done this and that and that (9h 8m 10s)

16/03/2022
----------

JIRA-123: Done this and that (7h 54m 49s)
JIRA-432: Dooooooone this and that (9h 58m 10s)
JIRA-1500: Dooooooone this and that (18h 18m 10s)
]